### PR TITLE
Refactor preferences dialog to take advantage of evented settings methods

### LINKED ISSUE #2993: Return key in preferences doesn't correspond to "ok"
## 🐛 Bug

Right now when I have the preferences dialog open and press the return key it doesn't correspond to "ok" instead it asks me if i want to return to the default settings. I'd expected it to correspond to ok. We might also want the shading of the "ok" button to be slightly different to indicate that it is "in focus" when the dialog opens.

cc @ppwadhwa @goanpeca

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR removes a lot of the logic in Preferences dialog that is more or less duplicated in our pydantic models.
For instance, there's a lot of logic to compare values and reset values and stuff, where most of that can be accomplished by:
```py
settings = get_settings()
state_0 = settings.dict()
# change settings as you want

# restore settings at a given state in time:
settings.update(state_0)
```

In general, we should take advantage of our evented persistent objects.  If we have a widget that gives a view onto that object (like the preferences dialog onto the global settings), then downstream objects that are dependent on the settings should be listening to the _settings_ object, not the widget.  So this removes methods like `_reset_preference_states` and `_update_widget_states` which added a level of indirection between the source of truth (the settings) and the target views.

This also closes #2993 by allowing preference changes made while the window was open to stay, unless the "cancel" button is clicked, in which case it returns to what it was when the window was opened.  I added tests for the things that I think the preferences dialog should do... but since there is a fair amount of untested code being removed here (whose functionality I can't automatically test), I'd appreciate a careful review (cc @ppwadhwa).

~**edit**: one thing I know I'm missing is the "restart needed" dialog.  need to figure out when that needs to show~  I added a warning to the settings model itself, this will show the restart message regardless of where the change is made

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/_qt/widgets/qt_dims_slider.py [hunk 1]
```diff
@@ -72,11 +72,17 @@ def __init__(self, parent: QWidget, axis: int):
...
-
         self._fps = settings.application.playback_fps
+        settings.application.events.playback_fps.connect(
+            lambda e: setattr(self, 'fps', e.value)
+        )
+
         self._minframe = None
         self._maxframe = None
         self._loop_mode = settings.application.playback_mode
+        settings.application.events.playback_fps.connect(
```
Comment (Carreau) [inline napari/_qt/widgets/qt_dims_slider.py]:
```suggestion
        settings.application.events.playback_mode.connect(
```

Maybe ? probably not `_fps` though.

**Comment by Carreau** [comment]
Thought on harmonizing loop_mode/playback_mode ? I know it could be a breaking change in config but the earlier the better ?

**Review by goanpeca — APPROVED** [review]

File: napari/_qt/widgets/qt_dims_slider.py (hunk 1 above)
Comment (Carreau) [reply napari/_qt/widgets/qt_dims_slider.py]: `playback_mode` seem to work when using  #3201 and corresponding IPython fix.

Comment (tlambert03) [reply napari/_qt/widgets/qt_dims_slider.py]: thank you!  will have a look soon

**Comment by tlambert03** [comment]
> Thought on harmonizing loop_mode/playback_mode ? I know it could be a breaking change in config but the earlier the better ?

yeah, I kinda prefer loop mode... it got named playback_mode when added to the settings.  Id be inclined to rename it to `self._loop_mode = settings.application.loop_mode`, thoughts?

File: napari/_qt/widgets/qt_dims_slider.py (hunk 1 above)
Comment (tlambert03) [reply napari/_qt/widgets/qt_dims_slider.py]: thanks for catching this

**Comment by ppwadhwa** [comment]
Hi @tlambert03 ,  
I think the way you re-worked things looks much simpler.  It looks great!  

I did notice a couple of things that aren't working yet.
I'm not sure if you know about these already, but I'll list them in any case.  

1.  On the experimental tab, if I select the octree settings, the async option is supposed to be disabled.  It is not disabled right now.
2. For both the plugins tab and shortcuts tab, if I make a change, and then choose to restore defaults, the defaults are not reset on the plugins or shortcuts widgets until napari is closed and re-opened. 
3. For the plugins and keyboard shortcuts tabs, if I make a change and then click `cancel`, the change I made is not canceled--at least it doesn't appear to be in the widget.  If I close and re-open napari, the old settings are there.  
4. If I go to the plugins tab (or keyboard shortcuts tab), make a change, click cancel.  Go back to the preferences dialog, click cancel (with or without the change), you will get a `ValidationError`.

**Comment by tlambert03** [comment]
thanks for the close look!  I'll make some fixes and write some tests

**Review by sofroniewn — APPROVED** [review]
Looks like this is now good to go, will merge!

**Pull request merged**